### ISSUE TITLE
Suggestions for monitoring.db

### ISSUE THREAD

**Issue by TomGlanzman** [Parsl/parsl#870]
Yadu suggested that I start an issue containing feedback on the current contents of monitoring.db.  This is a really important aspect of Parsl, particularly for large production campaigns and it is already very useful!  Here are some initial comments/suggestions.  More may be forthcoming as I get more experience with Parsl.

I used sqlite3 to inspect the monitoring.db file after a few Parsl runs on Cori at NERSC.  Found in the file were four tables:

 - **resource** = running, periodic, time-stamped updates on each 'task' 
 - **status**   = log of task state changes
 - **workflow** = overall workflow status summary
 - **task**     = overall task status summary

Table-specific comments:

1. status: add a column for 'hostname', which is needed to inspect a running job
2. resource: psutil_process_cpu_percent always seems to read "0.0"
3. resource: psutil_process_status always reads "sleeping"
4. task: add a column for "stderr"
5. task: add a new user-defined "task name" (string) to help associate a running task with its task_id.  This might be defined as an option in the app decorator?
6. resource: this table can get quite large very fast!  Consider a way to sparsify older routine entries.
7. task: add column for the rundir, e.g. "runinfo/NNN" so one can: 1) find the workflow logs; and 2) associate the run_id with the run number.

  Thanks for your consideration.

**Comment by annawoodard** [comment]
Thanks for the feedback, @TomGlanzman!!!

Cross referencing #849 which is related.

**Comment by ZhuozhaoLi** [comment]
Thanks for the comments! A few questions about your comments:

> status: add a column for 'hostname', which is needed to inspect a running job
do you mean the hostname that run the task?

>task: add column for the rundir, e.g. "runinfo/NNN" so one can: 1) find the workflow logs; and 2) associate the run_id with the run number.
I think you can locate the rundir in the workflow table, via the primary id `run_id`. For example, something like joining the tables:
`select workflow.rundir, workflow.run_id, task.task_id from workflow, task where workflow.run_id = task.run_id and task.task_id = 10;`

**Comment by TomGlanzman** [comment]
Hi @ZhuozhaoLi , 
   Yes, the hostname of the running task, e.g. (at NERSC) nid00609.  There are times when a user app misbehaves or the node upon which it is running misbehaves and one needs to log into that node to check/debug/kill the task.

  And thanks for the tip on getting the rundir.  Somehow I completely overlooked that column.  :)

**Comment by TomGlanzman** [comment]
It appears that the "workflow_version" column in the "workflow" table is being set to "time_began".   Is this correct?  What exactly is intended for "workflow_version"?

**Comment by ZhuozhaoLi** [comment]
Yes, it is intended to default the `workflow_version` as the `time_began`. `workflow_version` is actually used to separate different runs of the same workflow. You can manually set it in `MonitoringHub`: `MonitoringHub(workflow_version='0.1.0')`

**Comment by annawoodard** [comment]
Noting more issues-- I don't like monolithic issues, but I think it makes sense in this case, since many of these tweaks are so small. We can submit PRs in partial fulfillment of this issue.

- [ ] workflow summary titles should be human readable (no underscores)
- [ ] workflow duration should have sensible units and less precision

**Comment by ZhuozhaoLi** [comment]
> resource: psutil_process_cpu_percent always seems to read "0.0"
This seems fine on my test. I saw 100% reads with stress test. My guess is that @TomGlanzman 's apps probably have very low CPU utilization.

**Comment by ZhuozhaoLi** [comment]
>  resource: psutil_process_status always reads "sleeping"

This also works fine for me. I can see `running` state with stress test. 

Besides, even with stress test, I can still see some `sleeping` logs. So I would guess low-utilization apps would have `sleeping` very often.

**Comment by TomGlanzman** [comment]
Hi @ZhuozhaoLi   I ran that test on Cori-Haswell at NERSC.  The apps were *very* cpu intensive, however in this test I only ran on a small fraction of the total nodes.  Could it be that you are reporting total node CPU utilization rather than process utilization?  Just a thought...

**Comment by ZhuozhaoLi** [comment]
> Hi @ZhuozhaoLi I ran that test on Cori-Haswell at NERSC. The apps were _very_ cpu intensive, however in this test I only ran on a small fraction of the total nodes. Could it be that you are reporting total node CPU utilization rather than process utilization? Just a thought...

That is strange if all your apps were cpu-intensive. But maybe one possibility: the first data point of `CPU_percent` for each task is always 0.0 (`psutil`'s feature). Were your apps too short to have the second data point?

**Comment by TomGlanzman** [comment]
Hi @ZhuozhaoLi , My test job(s) ran for about 1.5 hours using the HTEX on Cori-Haswell with a single node.  I think that the monitoring.db file for that test is long gone.  Have you tried running a test on Cori-Haswell with HTEX?

**Comment by TomGlanzman** [comment]
Hi @ZhuozhaoLi , I found a monitoring.db file that does indicate the two issues noted above.  It is somewhat complicated due to the fact that this monitoring database has collected data from a range of successful and unsuccessful tests.  Let me know if this would help and I'll find a way to send it to you.

**Comment by ZhuozhaoLi** [comment]
Hi @TomGlanzman , thanks so much. It would be great if you could send that db file to me via slack.

**Comment by yadudoc** [comment]
> status: add a column for 'hostname', which is needed to inspect a running job

Done
 
> resource: psutil_process_cpu_percent always seems to read "0.0"
> resource: psutil_process_status always reads "sleeping"

I believe ZZ has not had success reproducing this issue. I'm going to try running a test on Cori.

> task: add a column for "stderr"

Split out to #987 

> task: add a new user-defined "task name" (string) to help associate a running task with its task_id. This might be defined as an option in the app decorator?

I believe #970 addresses this issue with a new `label` special kwargs to the function signature.

> resource: this table can get quite large very fast! Consider a way to sparsify older routine entries.

For now, `resource_monitoring_interval` is your best bet at rate limiting the collection of data. In the future we could try something like incrementally slowing down data collection. I don't think trimming a db is a good idea.

> task: add column for the rundir, e.g. "runinfo/NNN" so one can: 1) find the workflow logs; and 2) associate the run_id with the run number.

Done

**Comment by TomGlanzman** [comment]
Hi @yadudoc , with regard to the size of the monitoring.db file and the resource table in particular, one idea is to progressively sparsify the older data by reducing the frequency of saved data as one goes back in time but leaving the most recent data untouched.  Just an idea.  

What is your concern about trimming a database?  Is there some fundamental issue with doing so with a sqlite3 db file?

**Comment by yadudoc** [comment]
From the Parsl perspective the monitoring db is a log. I'm queezy about editing a log mid-run. I don't mind having a separate tool that is user invoked post-run that trims the db.

**Comment by TomGlanzman** [comment]
Another thought on the "sleeping"  and "CPU percent" issue.  

The bash_app I am running in my tests is defined like this:
```
@bash_app(executors=['coriHlocal'])
def genBFh(cmd, stdout='stdout.log', stderr='stderr.log'):
    return f'{cmd}'
```
Which is then invoked:
```
mycmd = '/a/b/c/myscript.sh'
myfuture = genBFh(mycmd)
```
Next, myscript.sh does a bit of setup then calls the LSSTDESC code to perform the heavy lifting:
```
/usr/bin/time -v /very/long/path/makeBrighterFatterKernel.py
```

Because there is a chain of sub-processes involved, could it be that *only* the top-level user process is assessed for "state" and "CPU percent" rather than (as I think it should be) the top-level user process **and** all of its sub-processes rolled up together?

**Comment by benclifford** [comment]
I also see the 0% cpu, `sleeping` behavior mentioned above, on some code that I am running.

1. My previous experience with monitoring systems is that recording strictly increasing counters (time and psutil_process_time_user, ..._system) and performing difference calculations at time of display works better than recording the difference calculations performed by the OS - and so the cpu percentage field should really be avoided for monitoring work (and probably the instantaneous state too for a lot of things)

2. The monitoring loop in `monitoring.py` does not call `pm.cpu_percent()` except at the start of the monitoring loop (where it is defined to return 0. My reading of https://psutil.readthedocs.io/en/latest/#processes suggests that it should be called in each iteration of the loop. I'm unclear how that behaviour matches up with @ZhuozhaoLi seeing a 100% CPU load as the parsl code as I read it can only ever return 0 for this value - but I could be misunderstanding this bit of code.

3. It looks like the task status is reported for the python process that runs the monitoring wrapper (that is, the process that is monitored is the process returned by `os.getpid()`. In the case of a bash app, this process probably really is doing nothing but wait for the child bash app process to exit. In the case of a python app, that process might actually be doing something with the CPU.

**Comment by ZhuozhaoLi** [comment]
>1. 

Agreed. The `cpu_user_time`, which is cumulative, seems a better metric for cpu usage. 

> 2. The monitoring loop in monitoring.py does not call pm.cpu_percent() except at the start of the monitoring loop

I believe the `cpu_percent` is called in this line.
https://github.com/Parsl/parsl/blob/da4702dda7e6f0bf9ff783f21f36b52208a27905/parsl/monitoring/monitoring.py#L413

> 3. It looks like the task status is reported for the python process that launches the monitoring wrapper (that is, the process that is monitored is the process returned by os.getpid() at the point the wrapper is launched). In the case of a bash app, this process probably really is doing nothing but wait for the child bash app process to exit. In the case of a python app, that process might actually be doing some application work with the CPU.

I think the psutil monitors the specific worker process: https://github.com/Parsl/parsl/blob/da4702dda7e6f0bf9ff783f21f36b52208a27905/parsl/monitoring/monitoring.py#L406
I think, for Python app, the task status is about the worker process -- it seems to get the right behavior
But for Bash app, the task status does not report properly. 

https://psutil.readthedocs.io/en/latest/#psutil.Process.status

**Comment by benclifford** [comment]
That `pid` comes from this invocation:

```
 def wrapped(*args, **kwargs):
            p = Process(target=monitor, args=(os.getpid()
```

in the monitor wrapper - which is the process running the python code, right? Not the actual bash application that is launched elsewhere by this:

```
        proc = subprocess.Popen(executable, stdout=std_out, stderr=std_err, shell=True, executable='/bin/bash')
``` in `parsl/app/bash.py`.

**Comment by yadudoc** [comment]
Migrating pending issues on this to `0.8.1`
